tutorials/realm/Barrier.md

```diff
@@ -0,0 +1,178 @@
+In this example, there are two types of tasks, reader and writer. 
+During each iteration, every writer task generates an integer value. Subsequently, 
+the collective sum of all these integers is calculated, and the resulting sum is then accessed by the reader tasks. 
+Thus, the writer and reader tasks take turns to execute.
+We illustrate how to use `Barrier` to go back and forth between 
+reader and writer tasks of different iterations.
+
+Here is a list of covered topics:
+
+* [Creating Barriers](#creating-barriers)
+* [Creating Reductions](#creating-reductions)
+* [Synchronizing Tasks with Barriers](#synchronizing-tasks-with-barriers)
+* [Performance Considerations](#performance-considerations)
+* [Limitations](#limitations)
+
+## Creating Barriers
+
+As introduced in the first section, each writer task performs a reduction operation on its local integer value. 
+The reader tasks can not read the reduction value until all writer tasks have completed writing. 
+This kind of synchronization can be implemented by `Barrier`. 
+In this example, we need to create two barriers, a `writer_barrier` and `reader_barrier`,
+to synchronize writer and reader tasks, respectively. 
+To create a barrier, we need to pass the number of concurrent barrier tasks into the `create_barrier` function.
+The following code demonstrates creating a `reader_barrier`, where we need to pass the number of participants:
+```c++
+Barrier reader_barrier = Barrier::create_barrier(TestConfig::num_readers);
+```
+
+## Creating Reductions
+
+The summation operation of writer tasks can be implemented using a reduction,
+which is declared as `ReductionOpIntAdd`. A reduction operator requires two functions, 
+`apply` combines a left-hand side type with a right-hand side type into a new left-hand side type, and `fold`
+combines two right-hand side types into a new right-hand side type, as well as an `identity`, a unique integer
+for Realm to recognize the reduction. The reduction operator needs to be registered by calling `register_reduction`
+before launching the main task, shown as follows:
+```c++
+rt.register_reduction<ReductionOpIntAdd>(REDOP_ADD);
+```
+
+The communication pattern of barrier and all reduction are similar; therefore, in the MPI world, these two
+routines have similar implementation. 
+Inspired by this, Realm extends the barrier to perform all-reduction-like computations, where each task can pass a number to
+the barrier. When reaching the synchronization point, the reduction value of all numbers is available to tasks that
+hold the barrier. 
+To perform reduction with a barrier, we can attach a reduction operator to the `create_barrier` function.
+The `init_value` tells the initial value of the parameter that the reduction operation is performed on. 
+```c++
+Barrier writer_barrier = Barrier::create_barrier(TestConfig::num_writers, REDOP_ADD,
+                                                 &init_value, sizeof(init_value));
+``` 
```

Comment:
And similarly, add a blank line after each code block...

Reply:
fixed for all realm tutorials